### PR TITLE
Updates to support ember-giftwrap

### PR DESCRIPTION
**Pull request by ef4**
This PR has two small changes:

1. A small refactor of the app tree broccoli pipeline. I improved it by doing the ES6 module conversion slightly earlier, in a place which is actually more logical, avoiding the need to merge the `_processedEmberCLITree` twice.

2. It makes the builder tasks's broccoli tree overridable.

Both of these changes are intended to support [Giftwrap](https://github.com/ef4/ember-giftwrap), which is a new general-purpose tool for packaging arbitrary sets of Ember addons for use by non-ember-cli apps. Code review of Giftwrap would also be appreciated.

### CODE REVIEW

**Comment by stefanpenner** [comment]
tests may be fixed by: https://github.com/ember-cli/ember-cli/pull/3811

**Comment by stefanpenner** [comment]
@rwjblue r?

File: lib/broccoli/ember-app.js [hunk 1]
```diff
@@ -818,8 +818,8 @@ EmberApp.prototype.appAndDependencies = function() {
   });
 
   var postprocessedApp = this.addonPostprocessTree('js', preprocessedApp);
+
   var sourceTrees = [
-    external,
```
Comment (ef4) [inline lib/broccoli/ember-app.js]: Needs review: I'm assuming here that the external tree isn't going to provide anything that matches the regexp used by `_es6ifyAppModules` below.

File: lib/broccoli/ember-app.js [hunk 2]
```diff
@@ -827,16 +827,36 @@ EmberApp.prototype.appAndDependencies = function() {
...
   this._addAppTests(sourceTrees);
 
-  var emberCLITree = this._processedEmberCLITree();
+  var es6app = this._es6ifyAppModules(mergeTrees(sourceTrees, {
+    overwrite: true,
+    description: 'TreeMerger (appModules)'
+  }));
 
-  sourceTrees.push(emberCLITree);
+  var emberCLITree = this._processedEmberCLITree();
```
Comment (ef4) [inline lib/broccoli/ember-app.js]: Likewise, I'm assuming that the processedEmberCLITree doesn't need to go through `_es6ifyAppModules`.

**Comment by ef4** [comment]
I don't know why only the iojs tests are reliably failing. It doesn't look related to me.

**Comment by stefanpenner** [comment]
> I don't know why only the iojs tests are reliably failing. It doesn't look related to me.
> Hide all checks

Ya, I believe its just an unrelated race, we need to dig into.

**Comment by trabus** [comment]
@rwjblue you okay with merging this in for 0.2.5?

**Comment by stefanpenner** [comment]
@ef4 is this something you still need for giftwrap?

**Comment by stefanpenner** [comment]
This seems to have gone stale, we can revisit. But an idle PR is just noise on the tracker.

**Comment by joostdevries** [comment]
@stefanpenner @ef4 Can I do anything to get this merged? I'm currently using giftwrap and it'd be better if it wouldn't need to depend on a forked ember-cli.

**Comment by rwjblue** [comment]
@joostdevries - I'm not sure if this is needed any longer, and ember-giftwrap doesn't list a custom/fork of ember-cli as a requirement.

**Comment by joostdevries** [comment]
@rwjblue https://github.com/ef4/ember-giftwrap/blob/master/package.json#L29 am I missing something?

**Comment by rwjblue** [comment]
Well, I feel dumb :stuck_out_tongue_closed_eyes:.  We'd need to track down the reasoning here, and see how to address in master.  Things have changed quite a lot since this PR was submitted....

**Comment by joostdevries** [comment]
Cool, I'll look into it day after tomorrow. Working on an Lambda-based addon giftwrapper (so we can use addons in Ember Twiddle) and since I need to bundle all node modules on every deploy I'm currently uploading 100MB to S3 for every test run :cry:...

**Comment by rwjblue** [comment]
@joostdevries - You should chat with @tomdale about his lamda experiences.  Also, as an aside, I personally think that it would be better to make a generic thing like I've done in https://github.com/emberjs/ember-collection/pull/92 (which is the "core" ember-giftwrap concept), and allow folks to provide these in their own S3 buckets...

**Comment by joostdevries** [comment]
I think the two should probably live side-by-side. Uploading your own app to s3 is good but quickly testing an addon without having to build it would be really awesome too. Got a POC working now (https://github.com/joostdevries/twiddle-backend). Might eventually actually run this on ECS instead of lambda though..

**Comment by rwjblue** [comment]
@joostdevries - I am saying that addon authors publish their addon for twiddle consumption to S3, not that app developers use S3.  See the [demo twiddle](https://ember-twiddle.com/4556826aac0e7b2a60ef?numColumns=1&openFiles=twiddle.json%2C) that I published with that PR.  We could easily add an additional key to the `twiddle.json` format, for `addons: ['ember-collection']`.

**Comment by joostdevries** [comment]
@rwjblue I understand but my aim is to make this something that just works(tm) for the majority of addons.

**Comment by tomdale** [comment]
@joostdevries You may want to check out https://github.com/tomdale/lambda-packager, a tool I used for compiling npm dependencies on AWS. It's a WIP but the idea is to upload your package.json to an EC2 instance where it runs `npm install` and uploads the resulting `node_modules` directory to S3. (This also helps if you have native dependencies like `contextify` in the case of FastBoot.)

I think you could prettily easily modify it so the pipeline would be something like:

```
+----------------------+         +----------------------+        +--------------------+
|                      |         |                      |        |                    |
|  Upload App          |  +----> | Install node_modules | +----> | Lambda             |
|  (no node_modules)   |         | into app directory   |        |                    |
|                      |         |                      |        |                    |
+----------------------+         +----------------------+        +--------------------+
   Dev machine                    Packager server (EC2)
```

TLDR have an intermediary build & deploy server that does the job of building `node_modules`. Because intra-AWS S3 usage is free it should considerably cut down both time and cost.

**Comment by joostdevries** [comment]
@tomdale awesome, will try it out
